File: Annchienta/doc/doxysrc/fake_sources/videomanager.py

## \brief Obtain the VideoManager instance.
#
#  Use this function to get access to the VideoManager
#  instance anywhere.
#
#  \return The global VideoManager instance.
def getVideoManager():
    pass


## \brief Handles video tasks.
#
#  This class is used for drawing, flipping and other video
#  operations. Remember to use setVideoMode() before actually
#  using it.
class VideoManager:

    ## \brief Sets the video mode.
    #
    #  Sets the current video mode as being screenWidth*screenHeight pixels.
    #
    #  \param screenWidth The screen width.
    #  \param screenHeight The screen height.
    #  \param title The window title.
    #  \param fullscreen Whether the window should be fullscreen'd or not.
    #
    def setVideoMode( screenWidth, screenHeight, title="Annchienta RPG Engine", fullscreen=False ):
        pass

    ## \brief Acquire the screen width.
    #
    #  \return The screen width.
    def getScreenWidth():
        pass

    ## \brief Acquire the screen height.
    #
    #  \return The screen height.
    def getScreenHeight():
        pass

    ## \brief Resets stuff.
    #
    #  This function resets the current matrix and color.
    def reset():
        pass

    ## \brief Translates the current matrix.
    #
    #  \param x X translate distance.
    #  \param y Y translate distance.
    def translate( x, y ):
        pass

    ## \brief Rotates the current matrix.
    #
    #  \param degrees Degrees to be rotated. Clockwise.
    def rotate( degrees ):
        pass

    ## \brief Scales the current matrix.
    #
    #  \param x X scale factor.
    #  \param y Y scale factor.
    def scale( x, y ):
        pass

    ## \brief Pushes the current matrix to the stack.
    #
    def pushMatrix():
        pass

    ## \brief Pops a matrix from the stack.
    #
    def popMatrix():
        pass

    ## \brief Begins the scene.
    #
    #  Clears and resets the scene.
    def begin():
        pass

    ## \brief Draws the buffer to the screen.
    #
    #  Ends the scene and draws the buffer to the screen.
    def end():
        pass

    ## \brief Sets the color.
    #
    #  Sets the color, used for surface, text and primitives drawing.
    #
    #  \param r Red Component. [0-255].
    #  \param g Green Component. [0-255].
    #  \param b Blue Component. [0-255].
    #  \param a Alpha Component. [0-255].
    def setColor( r, g, b, a ):
        pass

    ## \brief Sets the alpha component.
    #
    #  Sets the alpha component, used for surface, text and primitives drawing.
    #
    #  \param a Alpha Component. [0-255].
    def setAlpha( a ):
        pass

    ## \brief Draws a line.
    #
    #  \param x1 X coord of point 1.
    #  \param y1 Y coord of point 1.
    #  \param x2 X coord of point 2.
    #  \param y2 Y coord of point 2.
    def drawLine( x1, y1, x2, y2 ):
        pass

    ## \brief Draws a triangle.
    #
    #  \param x1 X coord of point 1.
    #  \param y1 Y coord of point 1.
    #  \param x2 X coord of point 2.
    #  \param y2 Y coord of point 2.
    #  \param x3 X coord of point 3.
    #  \param y3 Y coord of point 3.
    def drawTriangle( x1, y1, x2, y2, x3, y3 ):
        pass

    ## \brief Draws an axis-aligned rectangle.
    #
    #  \param x1 Top-left X coordinate.
    #  \param y1 Top-left Y coordinate.
    #  \param x2 Bottom-right X coordinate.
    #  \param y2 Bottom-right Y coordinate
    def drawRectangle( x1, y1, x2, y2 ):
        pass

    ## \brief Draws a quad.
    #
    #  You should pass the coordinates counterclockwise.
    #
    #  \param x1 X coordinate of point 1.
    #  \param y1 Y coordinate of point 1.
    #  \param x2 X coordinate of point 2.
    #  \param y2 Y coordinate of point 2.
    #  \param x3 X coordinate of point 3.
    #  \param y3 Y coordinate of point 3.
    #  \param x4 X coordinate of point 4.
    #  \param y4 Y coordinate of point 4.
    def drawQuad( x1, y1, x2, y2, x3, y3, x4, y4 ):
        pass

    ## \brief Draws a Surface.
    #
    #  \param surface Surface to be drawn.
    #  \param x Top-left X coord.
    #  \param y Top-left Y coord.
    def drawSurface( surface, x, y ):
        pass

    ## \brief Draws a part of a Surface.
    #
    #  \param surface Surface to be drawn.
    #  \param dx Destination X coordinate.
    #  \param dy Destination Y coordinate.
    #  \param sx1 Top-left X coordinate of source rectangle in the surface.
    #  \param sy1 Top-left Y coordinate of source rectangle in the surface.
    #  \param sx2 Bottom-right X coordinate of source rectangle in the surface.
    #  \param sy2 Bottom-right Y coordinate of source rectangle in the surface.
    def drawSurface( surface, dx, dy, sx1, sy1, sx2, sy2 ):
        pass

    ## \brief Draws a Surface as a pattern.
    #
    #  This function fills the rectangle defined by the function parameters
    #  with the given Surface.
    #
    #  \param surface Surface to be drawn.
    #  \param x1 Top-left X coordinate.
    #  \param y1 Top-left Y coordinate.
    #  \param x2 Bottom-right X coordinate.
    #  \param y2 Bottom-right Y coordinate.
    def drawPattern( surface, x1, y1, x2, y2 ):
        pass

    ## \brief Draws a string.
    #
    #  \param font Font to be used.
    #  \param string String to be drawn.
    #  \param x Upper-left X coord.
    #  \param y Upper-left Y coord.
    def drawString( font, string, x, y ):
        pass

    ## \brief Draws a string centered.
    #
    #  \param font Font to be used.
    #  \param string String to be drawn.
    #  \param x Middle X coord.
    #  \param y Upper-left Y coord.
    def drawStringCentered( font, string, x, y ):
        pass

    ## \brief Draws a string right-aligned.
    #
    #  \param font Font to be used.
    #  \param string String to be drawn.
    #  \param x Right X coord.
    #  \param y Upper Y coord.
    def drawStringRight( font, string, x, y ):
        pass

    ## \brief Grabs the buffer.
    #
    #  Copies the buffer to the given Surface.
    #  \attention The given Surface must be large enough to hold
    #  the buffer you are copying to it.
    #
    #  \param surface The destination Surface.
    def grabBuffer( surface ):
        pass

    ## \brief Grabs a part of the buffer.
    #
    #  Copies the rectangular area given by the coordinates to a given Surface.
    #
    #  \code
    # surface = annchienta.Surface( 50, 50 )
    # videoManager.grabBuffer( surface, 100, 100, 150, 150 )
    #  \endcode
    #
    #  \param surface The destination Surface.
    #  \param x1 Top-left X coord.
    #  \param y1 Top-left Y coord.
    #  \param x2 Bottom-right X coord.
    #  \param y2 Bottom-right Y coord.
    def grabBuffer( surface, x1, y1, x2, y2 ):
        pass

    ## \brief Saves the current buffer in a slot.
    #
    #  \param slot Number of the slot to be used. All implementations support at least slots [0-7].
    def storeBuffer( slot ):
        pass

    ## \brief Restores a buffer from a slot.
    #
    #  This basically draws a previsouly saved buffer to the screen.
    #
    #  \param slot Slot of the buffer which you would like to be restored.
    def restoreBuffer( slot ):

File: Annchienta/src/videomanager.cpp
/* This file is part of the Annchienta Project
 * Please consult the license and copyright details in Annchienta/license.txt
 */

#include "videomanager.h"

#include <SDL.h>
#include <GL/gl.h>

#include "surface.h"
#include "font.h"
#include "auxfunc.h"

namespace Annchienta
{
    VideoManager *videoManager;

    VideoManager::VideoManager() : screenWidth(0), screenHeight(0)
    {
        videoManager = this;

        reset();

        backBuffers = new Surface*[numberOfBackBuffers];
        for( int i=0; i<numberOfBackBuffers; i++ )
            backBuffers[i] = 0;
    }

    VideoManager::~VideoManager()
    {
        for( int i=0; i<numberOfBackBuffers; i++ )
            if( backBuffers[i] )
                delete backBuffers[i];
        delete[] backBuffers;
    }
    
    void VideoManager::setVideoMode( int w, int h, const char *title, bool fullscreen )
    {
        /* Get some variables.
         */
        screenWidth = w;
        screenHeight = h;

        /* Choose *best* settings for BitsPerPixel
         */
        Uint32 bpp = SDL_GetVideoInfo()->vfmt->BitsPerPixel;

        /* Preferred video flags.
         */
        Uint32 flags = SDL_HWSURFACE | SDL_OPENGL | SDL_HWACCEL;
        if( fullscreen )
            flags |= SDL_FULLSCREEN;

        /* Set the video mode.
         */
        SDL_Surface *screen = SDL_SetVideoMode( w, h, bpp, flags );

        /* Set the window title.
         */
        SDL_WM_SetCaption( title, NULL );

        /* Make sure we're in the right matrix.
         */
        glMatrixMode( GL_PROJECTION );
        glOrtho( 0, screenWidth, screenHeight, 0, -1, 1 );
        glViewport( 0, 0, screenWidth, screenHeight );

        glMatrixMode( GL_MODELVIEW );

        /* Set some flags.
         */
        glEnable( GL_TEXTURE_2D );
        glEnable( GL_BLEND );
        glBlendFunc( GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA );
        glEnable( GL_CULL_FACE );
        glCullFace( GL_BACK );

        glClearColor( 0.388f, 0.694f, 0.706f, 0.0f );

        for( int i=0; i<numberOfBackBuffers; i++ )
        {
            if( backBuffers[i] )
                delete backBuffers[i];
            backBuffers[i] = new Surface( screenWidth, screenHeight );
        }
    }

    int VideoManager::getScreenWidth() const
    {
        return screenWidth;
    }

    int VideoManager::getScreenHeight() const
    {
        return screenHeight;
    }

    void VideoManager::reset()
    {
        /* Reset the matrix.
        */
        glLoadIdentity();
    
        /* Reset the color.
        */
        this->setColor();

        /* Reset some flags.
         */
        glEnable( GL_TEXTURE_2D );
        glEnable( GL_BLEND );
        glBlendFunc( GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA );
    }
    
    void VideoManager::translate( float x, float y ) const
    {
        glTranslatef( x, y, 0.0f );
    }

    void VideoManager::rotate( float degrees ) const
    {
        glRotatef( degrees, 0.0f, 0.0f, 1.0f );
    }

    void VideoManager::scale( float x, float y ) const
    {
        glScalef( x, y, 1.0f );
    }

    void VideoManager::pushMatrix() const
    {
        glPushMatrix();
    }

    void VideoManager::popMatrix() const
    {
        glPopMatrix();
    }

    void VideoManager::begin()
    {
        glClear( GL_COLOR_BUFFER_BIT );
        this->reset();
    }

    void VideoManager::end()
    {
        SDL_GL_SwapBuffers();
    }

    void VideoManager::setColor( int red, int green, int blue, int alpha ) const
    {
        glColor4ub( red, green, blue, alpha );
    }

    void VideoManager::setAlpha( int alpha ) const
    {
        GLfloat colors[4];
        glGetFloatv( GL_CURRENT_COLOR, colors );
        colors[3] = (float)alpha/255.0f;
        glColor4fv( colors );
    }

    void VideoManager::setClippingRectangle( int x1, int y1, int x2, int y2 ) const
    {
        glScissor( x1, getScreenHeight()-y2, x2-x1, y2-y1 );
        glEnable( GL_SCISSOR_TEST );
    }

    void VideoManager::disableClipping() const
    {
        glDisable( GL_SCISSOR_TEST );
    }

    void VideoManager::drawLine( int x1, int y1, int x2, int y2 ) const
    {
        glDisable( GL_TEXTURE_2D );

        glBegin( GL_LINES );
            glVertex2f( x1, y1 );
            glVertex2f( x2, y2 );
        glEnd();

        glEnable( GL_TEXTURE_2D );
    }

    void VideoManager::drawTriangle( int x1, int y1, int x2, int y2, int x3, int y3 ) const
    {
        glDisable( GL_TEXTURE_2D );

        glBegin( GL_TRIANGLES );
            glVertex2f( x1, y1 );
            glVertex2f( x2, y2 );
            glVertex2f( x3, y3 );
        glEnd();

        glEnable( GL_TEXTURE_2D );
    }

    void VideoManager::drawRectangle( int x1, int y1, int x2, int y2 ) const
    {
        glDisable( GL_TEXTURE_2D );

        glBegin( GL_QUADS );
            glVertex2f( x1, y1 );
            glVertex2f( x1, y2 );
            glVertex2f( x2, y2 );
            glVertex2f( x2, y1 );
        glEnd();

        glEnable( GL_TEXTURE_2D );
    }

    void VideoManager::drawQuad( int x1, int y1, int x2, int y2, int x3, int y3, int x4, int y4 ) const
    {
        glDisable( GL_TEXTURE_2D );

        glBegin( GL_QUADS );
            glVertex2f( x1, y1 );
            glVertex2f( x2, y2 );
            glVertex2f( x3, y3 );
            glVertex2f( x4, y4 );
        glEnd();

        glEnable( GL_TEXTURE_2D );
    }

    void VideoManager::drawSurface( Surface *surface, int x, int y ) const
    {
        surface->draw( x, y );
    }

    void VideoManager::drawSurface( Surface *surface, int dx, int dy, int sx1, int sy1, int sx2, int sy2 ) const
    {
        float left = (float)sx1/(float)surface->getGlWidth(),
              right = (float)sx2/(float)surface->getGlWidth(),
              top = 1.0f - (float)sy1/(float)surface->getGlHeight(),
              bottom = 1.0f - (float)sy2/(float)surface->getGlHeight();

        glBindTexture( GL_TEXTURE_2D, surface->getTexture() );
        glBegin( GL_QUADS );

            glTexCoord2f( left, top );
            glVertex2f( dx, dy );

            glTexCoord2f( left, bottom );
            glVertex2f( dx, dy+sy2-sy1 );

            glTexCoord2f( right, bottom );
            glVertex2f( dx+sx2-sx1, dy+sy2-sy1 );

            glTexCoord2f( right, top );
            glVertex2f( dx+sx2-sx1, dy );
        glEnd();
    }

    void VideoManager::drawPattern( Surface *surface, int x1, int y1, int x2, int y2 ) const
    {
        int x, y;

        for( y = y1; y+surface->getHeight()<=y2; y+=surface->getHeight() )
        {
            for( x = x1; x+surface->getWidth()<=x2; x+=surface->getWidth() )
                drawSurface( surface, x, y );

            if( x != x2 )
                drawSurface( surface, x, y, 0, 0, x2-x, surface->getHeight() );
        }

        if( y != y2 )
        {
            for( x = x1; x+surface->getWidth()<=x2; x+=surface->getWidth() )
                drawSurface( surface, x, y, 0, 0, surface->getWidth(), y2-y );

            if( x != x2 )
                drawSurface( surface, x, y, 0, 0, x2-x, y2-y );
        }
    }

    void VideoManager::drawString( Font *font, const char *str, int x, int y ) const
    {
        font->draw( str, x, y );
    }

    void VideoManager::drawStringCentered( Font *font, const char *str, int x, int y ) const
    {
        font->draw( str, x - font->getStringWidth( str )/2, y );
    }

    void VideoManager::drawStringRight( Font *font, const char *str, int x, int y ) const
    {
        font->draw( str, x - font->getStringWidth( str ), y );
    }

    void VideoManager::grabBuffer( Surface *surface ) const
    {
        glBindTexture( GL_TEXTURE_2D, surface->getTexture() );

        glCopyTexSubImage2D( GL_TEXTURE_2D, 0, 0, surface->getGlHeight()-surface->getHeight(), 0, 0, surface->getWidth(), surface->getHeight() );
    }

    void VideoManager::grabBuffer( Surface *surface, int x1, int y1, int x2, int y2 ) const
    {
        if( x1>x2 )
            swap<int>(x1, x2);
        if( y1>y2 )
            swap<int>(y1, y2);

        int width  = x2 - x1,
            height = y2 - y1;

        glBindTexture( GL_TEXTURE_2D, surface->getTexture() );

        glCopyTexSubImage2D( GL_TEXTURE_2D, 0, 0, surface->getGlHeight()-surface->getHeight(), x1, getScreenHeight()-surface->getHeight()-y1, width, height );
    }

    void VideoManager::storeBuffer( int slot )
    {
        if( slot>=0 && slot<numberOfBackBuffers )
            this->grabBuffer( backBuffers[slot] );
    }

    void VideoManager::restoreBuffer( int slot ) const
    {
        if( slot>=0 && slot<numberOfBackBuffers )
            backBuffers[slot]->draw( 0, 0 );
    }

    VideoManager *getVideoManager()
    {
        return videoManager;
    }

};
